packages/hooks/src/useStateRef.ts

```diff
@@ -0,0 +1,30 @@
+import { useRef, useState } from 'react';
+
+/**
+ *
+ * A combination of useState and useRef,
+ * returns the current state, a `setState` and a `getState` function.
+ *
+ * Use the `getState` function inside an event listener callbacks
+ * in order to avoid referencing a stale state
+ *
+ * @param initial
+ * @returns [state, setState, getState]
+ */
+export function useStateRef<T extends any>(
+  initial: T,
+): [T, (x: T) => void, () => T] {
+  const [state, _setState] = useState<T>(initial);
+  const ref = useRef<T>(state);
+
+  function setState(newVal: T): void {
+    _setState(newVal);
+    ref.current = newVal;
+  }
+
+  function getState(): T {
+    return ref.current;
+  }
+
+  return [state, setState, getState];
```

Comment:
If I'm understanding this correctly, sometimes state and getState will not be in sync?

Reply:
During a render they will be in sync, but if you want to reference `state` in an event handler, there's a chance it could be stale